gpu/matrix/coo_kernels.cu

```diff
@@ -0,0 +1,386 @@
+
+namespace gko {
+namespace kernels {
+namespace gpu {
+namespace coo {
+
+
+constexpr int default_block_size = 512;
+
+
+namespace {
+
+
+__forceinline__ __device__ static float atomic_add(float *addr, float val)
+{
+    return atomicAdd(addr, val);
+}
+
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION < 8000)) || \
+    (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 600))
+__forceinline__ __device__ static double atomic_add(double *addr, double val)
+{
+    double old = *addr, assumed;
+    do {
+        assumed = old;
+        old = __longlong_as_double(atomicCAS(
+            (unsigned long long int *)addr, __double_as_longlong(assumed),
+            __double_as_longlong(val + assumed)));
+    } while (assumed != old);
+
+    return old;
+}
+#else
+__forceinline__ __device__ static double atomic_add(double *addr, double val)
+{
+    return atomicAdd(addr, val);
+}
+#endif
+
+
+__forceinline__ __device__ static cuDoubleComplex atomic_add(
+    cuDoubleComplex *address, cuDoubleComplex val)
+{
+    // Seperate to real part and imag part
+    // real part
+    atomic_add(&(address->x), val.x);
+    // imag part
+    atomic_add(&(address->y), val.y);
+    return *address;
+}
+
+
+__forceinline__ __device__ static cuComplex atomic_add(cuComplex *address,
+                                                       cuComplex val)
+{
+    // Seperate to real part and imag part
+    // real part
+    atomic_add(&(address->x), val.x);
+    // imag part
+    atomic_add(&(address->y), val.y);
+    return *address;
+}
+
+
+__forceinline__ __device__ static thrust::complex<float> atomic_add(
+    thrust::complex<float> *address, thrust::complex<float> val)
+{
+    cuComplex *cuaddr = reinterpret_cast<cuComplex *>(address);
+    cuComplex *cuval = reinterpret_cast<cuComplex *>(&val);
+    atomic_add(cuaddr, *cuval);
+    return *address;
+}
+
+
+__forceinline__ __device__ static thrust::complex<double> atomic_add(
+    thrust::complex<double> *address, thrust::complex<double> val)
+{
+    cuDoubleComplex *cuaddr = reinterpret_cast<cuDoubleComplex *>(address);
+    cuDoubleComplex *cuval = reinterpret_cast<cuDoubleComplex *>(&val);
+    atomic_add(cuaddr, *cuval);
+    return *address;
+}
+
+
+template <typename ValueType, typename IndexType>
+__device__ __forceinline__ void segment_scan(IndexType *ind, ValueType *val,
+                                             bool *head)
+{
+    ValueType add_val;
+#pragma unroll
+    for (int i = 1; i < 32; i <<= 1) {
+        const IndexType add_ind = warp::shuffle_up(*ind, i);
+        add_val = zero<ValueType>();
+        if (threadIdx.x >= i && add_ind == *ind) {
+            add_val = *val;
+            if (i == 1) {
+                *head = false;
+            }
+        }
+        add_val = warp::shuffle_down(add_val, i);
+        if (threadIdx.x < 32 - i) {
+            *val += add_val;
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(128) void spmv_kernel(
```

Comment:
As this is the actual kernel, some documentation explaining the working would be nice.